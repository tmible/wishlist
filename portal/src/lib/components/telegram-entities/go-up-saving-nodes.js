/** @typedef {import('./parser.svelte').EntitiesTreeNode} EntitiesTreeNode */
/** @typedef {import('./parser.svelte').Breakpoint} Breakpoint */

/**
 * Переход к родителю узла дерева с удалением текущего, если у него нет потомков
 * @function stepUpDeletingEmptyChild
 * @param {EntitiesTreeNode} node Узел дерева
 * @returns {EntitiesTreeNode} Новый узел дерева
 */
const stepUpDeletingEmptyChild = (node) => {
  const indexToDelete = node.children.length === 0 ? node.index : null;
  if (indexToDelete !== null) {
    node.parent.children.splice(
      node.parent.children.findIndex(({ index }) => index === indexToDelete),
      1,
    );
  }
  return node.parent;
};

/**
 * Закрытие элемента разметки текста с сохранением всех действующих. Фактически переход вверх
 * по дереву до нужного узла с запоминанием пройденных узлов и их повтоорением уже на одном уровне
 * с закрытым элементом
 * @function goUpSavingNodes
 * @param {EntitiesTreeNode} initialNode Изначальный текущий узел дерева
 * @param {Breakpoint} breakpoint Закрывающая точка изменения разметки текста
 * @returns {EntitiesTreeNode} Новый текущий узел дерева
 */
export const goUpSavingNodes = (initialNode, breakpoint) => {
  let node = initialNode;
  const repeatChildren = [];
  while (node.type !== breakpoint.entity.type) {
    repeatChildren.push(node.type);
    node = stepUpDeletingEmptyChild(node);
  }
  node = stepUpDeletingEmptyChild(node);
  while (repeatChildren.length > 0) {
    node.children.push({
      index: node.children.length,
      type: repeatChildren.pop(),
      children: [],
      parent: node,
    });
    node = node.children.at(-1);
  }
  return node;
};
