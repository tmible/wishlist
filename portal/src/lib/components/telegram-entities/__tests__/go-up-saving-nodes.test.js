import { beforeEach, describe, expect, it } from 'vitest';
import { goUpSavingNodes } from '../go-up-saving-nodes.js';

const breakpoint = { entity: { type: 'type 1' } };

describe('goUpSavingNodes', () => {
  let tree;

  beforeEach(() => {
    tree = {
      parent: null,
      children: [],
    };
    tree.children.push({
      type: 'type 1',
      parent: tree,
      children: [ 'child' ],
    });
  });

  it('should just go one level up if node type is equal to breakpoint entity type', () => {
    const [ initialNode ] = tree.children;
    expect(goUpSavingNodes(initialNode, breakpoint)).toBe(initialNode.parent);
  });

  it('should delete node if it has no children', () => {
    const [ initialNode ] = tree.children;
    initialNode.children = [];
    goUpSavingNodes(initialNode, breakpoint);
    expect(tree).toMatchSnapshot();
  });

  describe('passing several nodes', () => {
    beforeEach(() => {
      tree.children[0].children = [{
        index: 0,
        type: 'type 2',
        parent: tree.children[0],
        children: [],
      }, {
        index: 1,
        type: 'type 2',
        parent: tree.children[0],
        children: [],
      }];
      tree.children[0].children[1].children = [{
        index: 0,
        type: 'type 2',
        parent: tree.children[0].children[1],
        children: [ 'child' ],
      }];
    });

    it(
      'should go up until node type won\'t equal breakpoint ' +
      'entity type and repeat all passed nodes',
      () => {
        const [ initialNode ] = tree.children[0].children[1].children;
        goUpSavingNodes(initialNode, breakpoint);
        expect(tree).toMatchSnapshot();
      },
    );

    it('should delete all passed nodes that don\'t have children', () => {
      const [ initialNode ] = tree.children[0].children[1].children;
      initialNode.children = [];
      goUpSavingNodes(initialNode, breakpoint);
      expect(tree).toMatchSnapshot();
    });
  });
});
