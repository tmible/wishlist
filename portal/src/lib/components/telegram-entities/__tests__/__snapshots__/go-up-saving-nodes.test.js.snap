// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`goUpSavingNodes > passing several nodes > should delete all passed nodes that don't have children 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "index": 0,
          "parent": [Circular],
          "type": "type 2",
        },
      ],
      "parent": [Circular],
      "type": "type 1",
    },
    {
      "children": [
        {
          "children": [],
          "index": 0,
          "parent": [Circular],
          "type": "type 2",
        },
      ],
      "index": 1,
      "parent": [Circular],
      "type": "type 2",
    },
  ],
  "parent": null,
}
`;

exports[`goUpSavingNodes > passing several nodes > should go up until node type won't equal breakpoint entity type and repeat all passed nodes 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "index": 0,
          "parent": [Circular],
          "type": "type 2",
        },
        {
          "children": [
            {
              "children": [
                "child",
              ],
              "index": 0,
              "parent": [Circular],
              "type": "type 2",
            },
          ],
          "index": 1,
          "parent": [Circular],
          "type": "type 2",
        },
      ],
      "parent": [Circular],
      "type": "type 1",
    },
    {
      "children": [
        {
          "children": [],
          "index": 0,
          "parent": [Circular],
          "type": "type 2",
        },
      ],
      "index": 1,
      "parent": [Circular],
      "type": "type 2",
    },
  ],
  "parent": null,
}
`;

exports[`goUpSavingNodes > should delete node if it has no children 1`] = `
{
  "children": [],
  "parent": null,
}
`;
